consensus/hotstuff/vote_aggregator.go

```diff
@@ -2,27 +2,39 @@ package hotstuff
 
 import (
 	"github.com/onflow/flow-go/consensus/hotstuff/model"
-	"github.com/onflow/flow-go/model/flow"
 )
 
-// VoteAggregator aggregates votes and produces quorum certificates.
+// createVote holds the data for signing a block. The constructor of this function should have the block
+// info already, taking block again as argument is to ensure it's called with the same block.
+// For consensus cluster, the createVote function will create a disposable stateful signer object
+// for producing the signature for a vote. After the the signature has been produced, the signer object will
+// be discarded.
+type createVote func(*model.Block) (*model.Vote, error)
+
+// VoteAggregator verifies and aggregates votes to build QC.
+// When enough votes have been collected, it builds a QC and send it to the EventLoop
+// VoteAggregator also detects protocol violation, including invalid votes, double voting etc, and
+// notifies a HotStuff consumer for slashing.
 type VoteAggregator interface {
 
-	// StorePendingVote is used to store a vote for a block for which we don't
-	// have a block yet.
-	StorePendingVote(vote *model.Vote) (bool, error)
+	// AddVote verifies and aggregates a vote.
+	// The voting block could either be known or unknown.
+	// If the voting block is unknown, the vote won't be processed until AddBlock is called with the block.
+	// This method can be called concurrently, votes will be queued and processed asynchronously.
+	AddVote(vote *model.Vote) error
 
-	// StoreVoteAndBuildQC will store a vote and build the QC related to the
-	// voted upon block if enough votes can be accumulated.
-	StoreVoteAndBuildQC(vote *model.Vote, block *model.Block) (*flow.QuorumCertificate, bool, error)
+	// AddBlock notifies the VoteAggregator about a known block so that it can start processing
+	// pending votes whose block was unknown.
+	// It also verifies the proposer vote of a block, and return whether the proposer signature is valid.
+	AddBlock(block *model.Block) (bool, error)
 
-	// StoreProposerVote stores the vote of the proposer of the block and is
-	// used to separate vote and block handling.
-	StoreProposerVote(vote *model.Vote) bool
+	// GetVoteCreator returns a createVote function for a given block
+	// The caller must ensure the block is a known block by calling AddBlock before.
+	GetVoteCreator(block *model.Block) (createVote, error)
```

Comment:
Thanks for the write, and code examples. It's much easier for me to understand your point.

I like your idea to split interface into smaller ones, and only expose the smallest set of API possible. 

We can have the `BlockSigner` interface, but it can't be the VoteCollector who implements it. 

To be specific, for the following code:
```
vote, err := bp.blockSigner.CreateVote(block)
if err != nil {
	return fmt.Errorf("could not generate vote: %w", err)
}
header.ProposerSig = vote.SigData
```
The blockSigner here can't be the vote collector. Because when creating a vote, we want the vote collector (the stateful crypto signer) to hold the signature of the vote. Since the block is a new one, the vote collector must be newly created, and after it's used for creating a vote, we also need to add this vote collector to the VoteCollectors map, in order for it to collect votes from other nodes. 

Therefore, when EventHandler ask the BlockProducer to produce a block, there will be 3 actions after the block has been prepared:
1. creating the vote collector
2. use the create vote collector to create a proposal vote
3. add the vote collector to the vote collectors map. 

And this is how these 3 steps are done:
```
// step 1: create a vote collector instance outside of the vote collectors map.
voteCreator, err := bp.voteAggregator.GetVoteCreator(block)
if err != nil {
  return fmt.Errorf("could not instantiate vote creator: %w", err)
}

// step 2: voteCreator will first create a proposal vote
// step 3: then add the vote collector to the vote collectors map. Note, the closure of this 
// function must have to have access to the vote collectors map, in order to performance the action.
// step 4: after adding the new vote collector to the vote collectors map, it returns the vote back to the caller. 
vote, err := voteCreator(block)
if err != nil {
  return fmt.Errorf("could not generate vote: %w", err)
}
header.ProposerSig = vote.SigData
```


This was my original thinking. But I like your idea to simplify the 3 step into one `CreateVote`, in that case the blockSigner is not the vote collector itself. It will be such a module:
```
type BlockSigner struct {
  factory VoteCollectorFactory
  collectors VoteCollectors
}

func (s *BlockSigner) CreateVote(block) (vote, error) {
  voteCollector := s.factory.MakeVoteCreator(block)
  vote, err := voteCollector.CreateVote(block)
  if err != nil {
    return err // if we failed to create vote for the block, the vote collector won't be added to the map
  }
  collectors.AddCollector(voteCollector)
  return vote, nil
}
```